src/s2client-fsharp/ProtobufConnection.fs

```diff
@@ -1,256 +1,168 @@
+            let rec innerLoop curPos = async {
+                let left = sendBuf.Array.Length - curPos
+                if left <= 0 then
+                    return Error SendMessageBufferTooSmall
+                else
+                    try
+                        let segment = System.ArraySegment(receiveBuf.Array, curPos, left)
+                        let! result = client.ReceiveAsync(segment, tok) |> Async.AwaitTask
+                        match result.MessageType, result.EndOfMessage with
+                        |WebSocketMessageType.Binary, false ->
+                            return! innerLoop (curPos + result.Count)
+                        |WebSocketMessageType.Binary, true -> return Ok (curPos + result.Count)
+                        |_ -> return Error ExpectedBinaryResponse
+                    with
+                    |ex -> return ex |> FailedToReceiveMessage |> Error
+            }
+
+            let parseFrom finalPos =
+                SC2APIProtocol.Response.Parser.ParseFrom(new System.IO.MemoryStream(receiveBuf.Array, 0, finalPos))
+
+            return!
+                innerLoop 0
+                |> Result.mapAsyncInput parseFrom
+        }
+
+        let getAgent client = 
+            let getResponse() =
+                readMessage client
+            MailboxProcessor.Start (fun inbox ->
+                let rec messageLoop() = async{
+                    let! msg = inbox.Receive()
+
+                    match msg with
+                    |SendRequest (req, replyChannel) ->
+                        let! resp =
+                            writeMessage client req
+                            |> Result.bindAsync getResponse
+
+                        replyChannel.Reply resp
+
+                    return! messageLoop()
+                }
+                messageLoop()
+            )
+
+        let agent = connectedSocket |> Result.map getAgent
+
+        let postSendRequest req (agent:MailboxProcessor<ClientRequest>) =
+            agent.PostAndAsyncReply (fun reply -> SendRequest (req, reply))
+
+        let sendRequest req = async{
+            return!
+                agent
+                |> Result.bindAsyncBinder (postSendRequest req)
+        }
+
+        member this.SendRequest = sendRequest
+
+    let connect address port timeout tok = async{
+        return Sc2Connection(address, port, timeout, tok)
+    }
 
-    let doActions actions (cl : Sc2Connection) = async {
-        let request = new SC2APIProtocol.Request()
-        request.Action <- new RequestAction()
-        for action in actions do
-            request.Action.Actions.Add(action:Action)
-        let! response = sendRequest cl request
-        let actionResponse = response.Action
-        checkNullAndWarnings response actionResponse
-        return actionResponse.Result, response.Status }
\ No newline at end of file
+    let private applyFieldCheckAndReturnFunction fieldCheck returnFunc (response:SC2APIProtocol.Response)  =
```

Comment:
Can we rename `fieldCheck` to `getResponseField` and `returnFunc` to `getResult` to match the naming below?